File: app/src/main/java/com/ternaryop/photoshelf/util/MRU.java
package com.ternaryop.photoshelf.util;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.ListIterator;
import java.util.Locale;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.text.TextUtils;

/**
 * Created by dave on 30/11/17.
 * Handle Most Recently Used (MRU) list
 */

public class MRU {
    private static final String ITEM_SEPARATOR = "\n";
    private final SharedPreferences preferences;
    private final String key;
    private final int maxSize;
    private List<String> mruList;

    public MRU(Context context, String key, int maxSize) {
        preferences = PreferenceManager.getDefaultSharedPreferences(context);
        this.key = key;
        this.maxSize = maxSize;
    }

    public void add(List<String> items) {
        if (items == null) {
            return;
        }

        // every item is added at top so iterate in reverse order
        final ListIterator<String> iterator = items.listIterator(items.size());
        while (iterator.hasPrevious()) {
            add(iterator.previous());
        }
    }

    public boolean add(String item) {
        if (item == null || item.trim().isEmpty()) {
            return false;
        }
        String lowerCaseItem = item.toLowerCase(Locale.US);
        List<String> list = getList();
        if (!list.remove(lowerCaseItem) && list.size() == maxSize) {
            list.remove(list.size() - 1);
        }
        list.add(0, lowerCaseItem);
        return true;
    }

    public boolean remove(String item) {
        return getList().remove(item.toLowerCase(Locale.US));
    }

    public void save() {
        preferences
                .edit()
                .putString(key, TextUtils.join(ITEM_SEPARATOR, mruList))
                .apply();
    }

    public List<String> getList() {
        if (mruList == null) {
            mruList = new ArrayList<>(Arrays.asList(TextUtils.split(preferences.getString(key, ""), ITEM_SEPARATOR)));
        }
        return mruList;
    }
}

File: app/src/main/java/com/ternaryop/photoshelf/util/mru/MRU.kt
package com.ternaryop.photoshelf.util.mru

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import java.util.Locale

/**
 * Created by dave on 30/11/17.
 * Handle Most Recently Used (MRU) list
 */

private const val ITEM_SEPARATOR = "\n"

class MRU(context: Context, private val key: String, private val maxSize: Int) {
    private val preferences: SharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
    private val _list: MutableList<String> by lazy {
        preferences.getString(key, null)?.split(ITEM_SEPARATOR)?.mapTo(mutableListOf()) { it }
            ?: mutableListOf<String>()
    }

    val list: List<String> get() = _list

    fun add(items: List<String>) {
        // every item is added at top so iterate in reverse order
        for (item in items.asReversed()) {
            add(item)
        }
    }

    fun add(item: String): Boolean {
        if (item.trim { it <= ' ' }.isEmpty()) {
            return false
        }
        val lowerCaseItem = item.toLowerCase(Locale.US)
        if (!_list.remove(lowerCaseItem) && list.size == maxSize) {
            _list.removeAt(_list.size - 1)
        }
        _list.add(0, lowerCaseItem)

        return true
    }

    fun remove(item: String): Boolean {
        return _list.remove(item.toLowerCase(Locale.US))
    }

    fun save() {
        preferences
                .edit()
                .putString(key, _list.joinToString(ITEM_SEPARATOR))
                .apply()
    }
}
